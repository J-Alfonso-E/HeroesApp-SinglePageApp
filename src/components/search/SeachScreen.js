import React, { useMemo } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { HeroCard } from '../Heroes/HeroCard';
import { useForm } from "../hooks/useForm";
import queryString from 'query-string';
import { getHeroesByName } from "../../selectors/getHeroesByName";

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const [formValues, handleInputChange] = useForm({ searchText: q });

    const { searchText } = formValues;
    //const HeroesFiltered = getHeroesByName(q);
    const HeroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    

    const EnviarFormulario = (e) => {
        e.preventDefault();
        navigate(`?q=${searchText}`);
    }

    //console.log(HeroesFiltered);
    return (
        <div>
            <h1>SearchScreen</h1>
            <hr />
            <div className="row">

                <div className="col-5">
                    <h4>Buscar</h4>
                    <hr />

                    <form onSubmit={EnviarFormulario}>
                        <input type="text" placeholder="Buscar un Heroe" className="form-control" name="searchText" autoComplete="off" value={searchText} onChange={handleInputChange} />
                        <button className="btn btn-primary mt-1" type="submit">Buscar</button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Resultado</h4>
                    
                    {
                        (q === "") ? <div className="alert alert-info">Buscar un heroe</div> : (HeroesFiltered.length === 0 ) && <div className="alert alert-danger">No se ha encontrado coincidencias con la busqueda: {q}</div>
                    }

                    {
                        HeroesFiltered.map(hero => (
                            <HeroCard key={hero.id} { ...hero} />
                        ))
                    }

                </div>
            </div>



        </div>);
}