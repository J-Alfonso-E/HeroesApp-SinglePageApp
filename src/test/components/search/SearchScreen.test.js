import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SeachScreen';

const mockNavigate = jest.fn(); //Funcion de prueba de useNavigate, forzosamente se requiere que comience con "mock"

jest.mock('react-router-dom', () => ({ //Simulacion de la libreria de de react-router-dom
    ...jest.requireActual('react-router-dom'), //Importa la libreria de react sin cambios
    useNavigate: () => mockNavigate //Despues de que se importo se sobreescribe el hook de useNavigate
}));


describe("Pruebas con el SearchScreen", () => {
    

    test('Snapshot vacio', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/seach']} >
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un heroe');
    });

    test('debe de mostrar la imagen de Batman y el campo de busqueda con batman', () => { 
        const wrapper = mount(
            <MemoryRouter initialEntries={['/seach?q=batman']} >
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
        
    });

    test('debe de mostrar una etiqueta de que no encontro el heroe', () => { 
        const wrapper = mount(
            <MemoryRouter initialEntries={['/seach?q=batman123']} >
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper.find('.alert-danger').text().trim()).toBe('No se ha encontrado coincidencias con la busqueda: batman123');
        expect(wrapper.find('input').prop('value')).toBe('batman123');
        
        
    });
});
