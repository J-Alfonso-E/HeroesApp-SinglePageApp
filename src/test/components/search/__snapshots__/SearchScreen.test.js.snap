// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con el SearchScreen Snapshot vacio 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un Heroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary mt-1"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultado
      </h4>
      <div
        className="alert alert-info"
      >
        Buscar un heroe
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas con el SearchScreen debe de mostrar la imagen de Batman y el campo de busqueda con batman 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un Heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary mt-1"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultado
      </h4>
      <div
        className="col-md-3 animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <img
              alt="Batman"
              className="card-img"
              src="/assets/dc-batman.jpg"
            />
          </div>
          <div
            className="col-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                Detective Comics #27
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                 Mas... 
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
